pymedphys/_experimental/pseudonymisation/strategy.py

```diff
@@ -0,0 +1,379 @@
+# Copyright (C) 2020 Stuart Swerdloff, Simon Biggs
+# Copyright (C) 2018 Matthew Jennings, Simon Biggs
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import base64
+import datetime
+import hashlib
+import logging
+import random
+from decimal import Decimal, DecimalTuple
+
+from pymedphys._imports import pydicom
+
+from pymedphys._dicom.anonymise import get_baseline_keyword_vr_dict
+from pymedphys._dicom.constants import PYMEDPHYS_ROOT_UID
+
+EPOCH_START = "20000101"
+DEFAULT_EARLIEST_STUDY_DATE = "20040415"
+DICOM_DATE_FORMAT_STR = "%Y%m%d"
+DICOM_DATETIME_FORMAT_STR = "%Y%m%d%H%M%S.%f"
+
+
+def get_pseudonymous_replacement_value(keyword, value):
+    """
+    Get an appropriate value for a DICOM element based on its VR and
+    take in to account if it is an SOP Class UID.
+    Cache values of keywords that have already come in and re-use
+    the pseudonym for the *value* if it has already been used/replaced.
+
+    Parameters
+    ----------
+    keyword : TYPE
+        DESCRIPTION.
+
+    Returns
+    -------
+    None.
+
+    """
+    vr = get_baseline_keyword_vr_dict()[keyword]
+    replacement_value = value
+    if vr in pseudonymisation_dispatch:
+        replacement_value = pseudonymisation_dispatch[vr](value)
+    else:
+        logging.warning("VR of %s not found in pseudonymisation dispatch table", vr)
+
+    return replacement_value
+
+
+def _pseudonymise_plaintext(value):
+    """
+    Appropriate for DICOM VR that are unstructured text, e.g. SH, AE, LO
+    Not appropriate for Numeric Strings (DS, IS)
+    Not appropriate for DateTime (DA, TM, DT)
+
+    Parameters
+    ----------
+    value : String
+        the DICOM element value for an element of type SH, AE, LO.
+
+    Returns
+    -------
+    The base64 text representation of the hashed input value
+
+    """
+    try:
+        encoded_value = value.encode("ASCII")
+    except AttributeError:
+        encoded_value = str(value).encode("ASCII")
+
+    my_hash_func = hashlib.new("sha3_256")
```

Comment:
Was the intent for the user to be able to revert these values if they had a private key? I thought the approach might involve a reversible encryption step.

Reply:
eg https://cryptography.io/en/latest/fernet/